### PR TITLE
exporting packages with features

### PR DESCRIPTION
**Pull request by dimbleby**
Fix for #5537, see also https://github.com/python-poetry/poetry-plugin-export/pull/65

Have `get_project_dependency_packages()` return packages with features suitably set, so that the exporter can record them.  Features are accumulated as we walk the tree.

With that set of fixes, [this test](https://github.com/python-poetry/poetry-plugin-export/blob/db1db238d521462960616463741f20c63c9cb6b7/tests/test_exporter.py#L2140-L2143) sees this diff
```patch
-localstack-ext==1.0.0 ; {MARKER_PY36}
-localstack==1.0.0 ; {MARKER_PY36}
+localstack-ext[bar]==1.0.0 ; {MARKER_PY36}
+localstack[foo]==1.0.0 ; {MARKER_PY36}
```
which is https://github.com/python-poetry/poetry-plugin-export/pull/66

Direct manipulation of `_extras` is not great, we should add `set_extras(extras)` or similar in `poetry-core`.  That's a little exciting - `_extras` gets hashed so careless callers can go wrong - but it's not worse than `set_constraint()` as also used by the locker (and, similarly, is safe to do in this code).

Also we can remove `requires_extras` from poetry-core, it's unused.

### CODE REVIEW

**Comment by dimbleby** [comment]
I don't know what order these things can all be committed in to keep the tests passing at all times!

File: src/poetry/packages/locker.py [hunk 1]
```diff
@@ -268,6 +268,7 @@ def __walk_dependencies(
             extras = requirement.extras
             requirement = locked_package.to_dependency()
             requirement.marker = requirement.marker.intersect(marker)
+            requirement._extras = extras
```
Comment (abn) [inline src/poetry/packages/locker.py:271]:
```suggestion
            requirement._extras = extras
            requirement._features = frozenset(extras)
```

In deference to the change proposed in https://github.com/python-poetry/poetry-core/pull/373/files#r882064718 this should allow us to update to next core without issues.

File: src/poetry/packages/locker.py [hunk 2]
```diff
@@ -283,12 +284,12 @@ def __walk_dependencies(
...
-            if key not in nested_dependencies:
-                nested_dependencies[key] = requirement
-            else:
-                nested_dependencies[key].marker = nested_dependencies[key].marker.union(
+            combined_requirement = nested_dependencies.setdefault(key, requirement)
+            if combined_requirement is not requirement:
+                combined_requirement.marker = combined_requirement.marker.union(
                     requirement.marker
                 )
+                combined_requirement._extras |= requirement.extras
```
Comment (abn) [inline src/poetry/packages/locker.py:292]: Similar as above.

**Review by abn — CHANGES_REQUESTED** [review]
Thanks @dimbleby for picking this up. Some comments in relation to fixing the duplicated attributes in core's `Dependency` implementation.

File: src/poetry/packages/locker.py (hunk 1 above)
Comment (dimbleby) [reply src/poetry/packages/locker.py:271]: The intention is that this MR fixes things with the current poetry-core, and when poetry-core exposes a better API we use it here - which I would expect to do with a later MR.

If you want to block this behind publishing poetry-core improvements I guess that's fine by me, then we can do it in a single commit here.

**Comment by abn** [comment]
One possible order for these changes could be as follows.

1. Add a ` || true` to the export test command in ci so the workflow does not bail
2. Relock the dev dependency in the export plugin in https://github.com/python-poetry/poetry-plugin-export/pull/66
3. Make a new release of the export plugin
4. Update plugin version on poetry and undo (1).

In the long term we need to fix this coupling somehow :(

**Comment by dimbleby** [comment]
Long-term, I suppose that this repository should simply not know anything about poetry-plugin-export; because that plugin should be no more special than any other.

(Also that would fix the circular dependency where both projects depend on one another).

Perhaps we can just do that now, must we test the export plugin from here?

File: src/poetry/packages/locker.py (hunk 1 above)
Comment (abn) [reply src/poetry/packages/locker.py:271]: I am simply anticipating a forward port required to get the downstream tests passing. And suggesting that we remove `requirement._extras = extras` once the new API is available.

I'd rather not block this behind a new core release and have this fixed in the next poetry pre-release that will likely happen prior to next core release.

**Comment by abn** [comment]
> Perhaps we can just do that now, must we test the export plugin from here?

It is possible yes. However, might require a deprecation cycle for the command itself. The issue is `export` command has been a core poetry command, and changes in poetry can break the export command. The tests were introduced because of one such case a while back where `poetry export` broke badly.

**Comment by dimbleby** [comment]
reworked fix is now simpler, and exports like this in that existing testcase:
```patch
 localstack-ext==1.0.0 ; {MARKER_PY36}
-localstack==1.0.0 ; {MARKER_PY36}
+localstack-ext[bar]==1.0.0 ; {MARKER_PY36}
+localstack[foo]==1.0.0 ; {MARKER_PY36}
```
ie we get both localstack-ext and localstack-ext[bar]

This is cosmetically ugly when, as here, both variations have the same markers. But in cases where they have different markers eg some extra is required on windows only - this seems like it is giving the right answers where the previous try didn't.

**Review by abn — APPROVED** [review]

**Pull request merged**